htlcswitch/link.go

```diff
@@ -0,0 +1,882 @@
+		// peer.
+		l.sendHTLCError(packet.payHash, htlc.Reason)
+		needUpdate = true
+	}
+
+	// If this newly added update exceeds the min batch size for adds, or
+	// this is a settle packet, then initiate an update.
+	// TODO(roasbeef): enforce max HTLCs in flight limit
+	if l.batchCounter >= 10 || needUpdate {
+		if err := l.updateCommitTx(false); err != nil {
+			l.cfg.Peer.Disconnect()
+			return errors.Errorf("channel link(%v): unable to "+
+				"update commitment: %v", l, err)
+		}
+	}
+
+	return nil
+}
+
+// Peer returns the representation of remote peer with which we
+// have the channel link opened.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Peer() Peer {
+	return l.cfg.Peer
+}
+
+// ChannelPoint returns the unique identificator of the channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) ChannelPoint() wire.OutPoint {
+	return *l.channel.ChannelPoint()
+}
+
+// Bandwidth returns the amount which current link might pass
+// through channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Bandwidth() btcutil.Amount {
+	snapshot := l.channel.StateSnapshot()
+	return snapshot.LocalBalance
+}
+
+// Stats return the statistics of channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Stats() (uint64, btcutil.Amount, btcutil.Amount) {
+	snapshot := l.channel.StateSnapshot()
+	return snapshot.NumUpdates,
+		btcutil.Amount(snapshot.TotalSatoshisSent),
+		btcutil.Amount(snapshot.TotalSatoshisReceived)
+}
+
+// String returns the string representation of channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) String() string {
+	return l.ChannelPoint().String()
+}
+
+// Start starts all helper goroutines required for the operation of the
+// channel link.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Start() error {
+	if !atomic.CompareAndSwapInt32(&l.started, 0, 1) {
+		log.Warn("channel link(%v): already started", l)
+		return nil
+	}
+
+	log.Info("channel link(%v): starting", l)
+
+	l.wg.Add(1)
+	go l.startHandle()
+
+	// A new session for this active channel has just started, therefore we
+	// need to send our initial revocation window to the remote peer.
+	for i := 0; i < lnwallet.InitialRevocationWindow; i++ {
+		revocation, err := l.channel.ExtendRevocationWindow()
+		if err != nil {
+			log.Errorf("channel link(%v): unable to expand "+
+				"revocation window: %v", l, err)
+			continue
+		}
+		l.cfg.Peer.SendMessage(revocation)
+	}
+
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+// NOTE: Part of the ChannelLink interface.
+func (l *channelLink) Stop() {
+	if !atomic.CompareAndSwapInt32(&l.shutdown, 0, 1) {
+		log.Warn("channel link(%v): already stopped", l)
+		return
+	}
+
+	log.Info("channel link(%v): stopping", l)
+
+	close(l.quit)
+}
+
+// Wait waits for service to stop.
```

Comment:
By merging `Wait` with `Stop` with we will hit the deadlock which I described in the description of `Wait` function.